Remove autoconfigured audience for GCE/GKE (#1293)
spring-cloud/spring-cloud-gcp: docs/src/main/asciidoc/security-iap.adoc

 * Issue time
 * Expiration time
 * Issuer
 * Audience

> The audience (`"aud"`) validation is automatically configured when the application is running on AppEngine.

Reviewer: ```suggestion
The audience (`"aud"`) validation is automatically configured when the application is running on App Engine Standard.
```

Author: It should work the same on AppEngine Standard and AppEngine Flexible.

For other runtime environments, a custom audience must be provided through `spring.cloud.gcp.security.iap.audience` property.

> The custom property, if available, overrides the automatic AppEngine check.

Reviewer: ```suggestion
The custom property, if specified, overrides the automatic App Engine audience detection.
```

IMPORTANT: There is no automatic audience string configuration for Compute Engine or Kubernetes Engine.

> To use the IAP starter on GCE/GKE, find the Audience string per instructions the https://cloud.google.com/iap/docs/signed-headers-howto#verify_the_jwt_payload[Verify the JWT payload] guide, and specify it in `spring.cloud.gcp.security.iap.audience` property.

Reviewer: ```suggestion
To use the IAP starter on GCE/GKE, find the Audience string per instructions in the https://cloud.google.com/iap/docs/signed-headers-howto#verify_the_jwt_payload[Verify the JWT payload] guide, and specify it in the `spring.cloud.gcp.security.iap.audience` property.
```

 * Expiration time
 * Issuer

Additionally, Audience (`"aud"`) validation is enabled when the application is running on AppEngine, or when a custom audience is provided through `spring.cloud.gcp.security.iap.audience` property.
The custom property, if available, overrides the automatic AppEngine check.

IMPORTANT: The audience check is *not* enabled by default on Compute Engine or Kubernetes Engine.

> To fully secure an application running on Compute Engine or Kubernetes Engine, find the Audience string per instructions the https://cloud.google.com/iap/docs/signed-headers-howto#verify_the_jwt_payload[Verify the JWT payload] guide, and specify it in `spring.cloud.gcp.security.iap.audience` property.

Reviewer: If not checking the audience is a security vulnerability, we should not allow this by default. It's a pity that we can't determine it automatically though.

Author: I am on the fence about this. On the one hand, this is more of a backup, defense-in-depth mechanism -- IAP proxy itself will validate all claims before routing to the application. On the other hand, if someone does figure out how to bypass IAP, not validating audience means one can get a [valid, recent, GCP-signed] IAP token from any application and it will be accepted.

Author: For posterity -- we are removing GCE/GKE audience autoconfiguration for now, and reaching out to IAP team for guidance.

 * Expiration time
 * Issuer

> Additionally, Audience (`"aud"`) validation is enabled when the application is running on AppEngine, or when a custom audience is provided through `spring.cloud.gcp.security.iap.audience` property.

Reviewer: Isn't audience validation always enabled? It's just that the audience is automatically detected on App Engine, but not on GCE/GKE.
